### ISSUE TITLE
Lazy evaluation in `let` block does not behave as expected

### ISSUE THREAD

**Issue by stefan-hoeck** [idris-lang/Idris2#1900]
It should be possible to delay evaluation in a `let` expression, in case we are dealing with an expensive computation, the result of which will not necessarily be needed. Right now, this doesn't seem to work (at least not when using `delay`, but probably I'm doing something wrong here).

# Steps to Reproduce

Build and run the following program:

```idris
import Debug.Trace

calc : Nat -> Nat
calc n0 =
  let n1  = delay $ (trace "foo" $ n0 + n0)
      n2  = delay $ (trace "bar" $ 2 * n1)
      n3  = delay $ (trace "baz" $ 2 * n2)
   in if n0 > 10 then n2 else n3

main : IO ()
main = do printLn (calc 100)
```

# Expected Behavior

Program prints

```
foo
bar
400
```

# Observed Behavior

Program prints

```
foo
bar
baz
400
```

It can be seen in the generated Javascript, that the call to `delay` is ignored (but this is not an issue of the JS backend. The schemes behave the same):

```javascript
function Main_calc($0) {
 const $1 = Debug_Trace_trace('foo', Prelude_Types_x2b_Num_Nat($0, $0));
 const $8 = Debug_Trace_trace('bar', Prelude_Types_x2a_Num_Nat(Prelude_Types_fromInteger_Num_Nat(2n), $1));
 const $11 = Debug_Trace_trace('baz', Prelude_Types_x2a_Num_Nat(Prelude_Types_fromInteger_Num_Nat(2n), $8));
 switch(Prelude_Types_x3e_Ord_Nat($0, Prelude_Types_fromInteger_Num_Nat(10n))) {
  case 1: return $8;
  case 0: return $11;
 }
}
```

**Comment by gallais** [comment]
Do you get the same behaviour if you explicitly assign a `Lazy ?` type to these
let-bound expressions? It could be that inference comes up with a type that means
the thunk is forced!

**Comment by stefan-hoeck** [comment]
You are right. If I replace `delay` with `the (Lazy Nat)`, I get the expected behavior.

**Comment by stefan-hoeck** [comment]
So, how to move on from here? Personally, I think the inferred type here is not correct, as the call to `delay` should make it clear that we are interested in a delayed computation. However, it might be difficult to convey this to Idris, given that at other occasions we are happy that Idris inserts `force` for us automatically. As an alternative, we could mention this use case somewhere in the documentation and accept it as one of the quirks of Idris2.

I wonder, however, if these silent conversions are really that helpful. Sure, they make it sometimes convenient to write lazy code. But  personally, I'm not very happy with them, as they more often than not make it hard to grasp (for me at least) what's going on.

**Comment by Z-snails** [comment]
How about a switch to disable implicit force and delay (like the one for disabling unbound implicits in type signatures)?

**Comment by stefan-hoeck** [comment]
> How about a switch to disable implicit force and delay (like the one for disabling unbound implicits in type signatures)?

I think such a thing would be very useful. Turning these implicits off when learning Idris2 might give programmers a good idea, where and when these conversions need to be inserted. Once they feel confident with the concept, they might still opt to turn the implicit conversions on.

**Comment by buzden** [comment]
> How about a switch to disable implicit force and delay (like the one for disabling unbound implicits in type signatures)?

I think that if to create such a switch, it should support a mode when only implicit `force` is turned off, but implicit `delay` would be on. It would enforce potentially costly operation to be explicit while leaving simplicity of writing delayed values.

**Comment by Z-snails** [comment]
@buzden I don't agree, because where the delay is put could cause expensive computations, eg
```
foo : Lazy Nat
foo =
    let x = expensive 10
     in x * 20
```
it's possibly ambigous wether that would be
```
foo =
    let x = expensive 10
     in delay (x * 20)
```
(runs the expensive computation immediately)
or
```
foo = delay $
    let x = expensive 10
     in x * 20
```
(runs the expensive computation only when it's needed)
